steps/24_dataset_object/step.md

```diff
@@ -0,0 +1,349 @@
+5. [Discussion and Comparison of Alternative Solutions](#discussion-and-comparison-of-alternative-solutions)
+6. [Detailed Description of Design and Implementation](#detailed-description-of-design-and-implementation-of-proposed-solution)  
+
+## Problem statement
+
+Currently, datasets in sktime are loaded by the usage of functions, and having a dataset object would provide many benefits for documentation purposes. For example, in the case of estimators, there's an estimator overview page that allows users to filter and find estimators with certain characteristics for classification, forecasting and regression.
+Having dataset objects would allow the retrieval of problems with characteristics that fit the user needs by the usage of the tag system.
+
+In addition to this, an unified interface that also identifies the characteristics of each dataset can be useful for benchmarking purposes. If a researcher is investigating in which circunstances a model performs well, the metadata (tags) of the datasets could be useful to analize and understand these factors.
+
+The dataset object interface could also be used for other scenarious beyond timeseries forecasting, classification and regression. For example, the forecasting dataset object could be used for Marketing Mix Modeling datasets (see felipeangelimvieira/prophetverse#101)
+
+## Description of proposed solution
+
+### `__init__` and `load`
+
+Dataset objects should have all the parameters that identify the content of it defined in `__init__`, similar to what hyperparameters represent for estimators. The `load(*args)` method is used to obtain the sets. This method receives strings as positional args, such as `load("X", "y", "X_train", "y_train", "X_test", "y_test")` and returns the respective sets. Hence, users can define which set should be retrieved by passing its name to `load`. 
+
+### Train and test splits
+
+Many public datasets have default train-tests splits, and could also potentially have default cross-validation splits. Others, however, do not provide such sets. The following behaviour will be used to handle this heterogeneity:
+
+* Calling load with "X_train", "X_test", "y_train" and "y_test" should raise `InvalidSetException` if the dataset does not have a pre-defined split. The sktime interface will not provide a custom split if the original dataset does not define one.
+* In the case of existence of a cross validation split, "cv" keyword can be used, and a generator with the CV folds will be returned. This generator will have length equal to the number of folds in the CV split, and contain the "X_train", "X_test", "y_train" and "y_test" data for each split. Datasets with a single train-test split should return generators of length one.
+* If the dataset has a CV split that is not a simple train-test split, then calling `"X_train"`, `"X_test"`, `"y_train"` and `"y_test"` should also raise `InvalidSetException`, and an informative error message recommending the user to call `"cv"`.
```

Comment:
alternative: returns a generator instead. Though I am not certain which of the options is better. Perhaps better to concentrate the valid API points for now.